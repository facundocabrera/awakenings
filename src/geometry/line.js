const { sqrt, cos, sin, round } = Math;

import {  
  angle_between,
  unit_vector,
  distance,
  direction
} from './vector';

/**
 * Dado un segmento definido por `(x1, y1)` => `(x2, y2)` computar los puntos
 * que entran saltando `step` unidades.
 *
 * @param step unidades que deseo avanzar
 * @param angle en RADIANES (si no lo pasamos, se calcula automaticamente en base a los puntos)
 */
const line_points = (x1, y1, x2, y2) => {
  const [ux1, uy1] = unit_vector(x1, y1);
  const [ux2, uy2] = unit_vector(x2, y2);

  const angle = angle_between(ux1, uy1, ux2, uy2);
  // teorema del coseno simplificado por el uso de vectores unitarios
  const step = sqrt(2 - 2 * cos(angle)); 

  const dx = x2 - x1 < 0 ? -1 : 1;
  const dy = y2 - y1 < 0 ? -1 : 1;

  const steps = round(distance(x1, y1, x2, y2) / step);

  let incx = step * cos(direction(x1, y1, x2, y2));
  let incy = step * sin(direction(x1, y1, x2, y2));

  // ðŸ§  ðŸ§  ðŸ§ 
  // Esta correcion de signo fue lo mejor que se me ocurrio para
  // se armen correctamente los puntos dentro de los extremos
  // que estoy pasando de parametro.
  if ((dx < 0 && incx > 0) || (dx > 0 && incx < 0)) {
    incx *= -1;
  }
  if ((dy < 0 && incy > 0) || (dy > 0 && incy < 0)) {
    incy *= -1;
  }

  const points = [];

  for (let i = 0; i <= steps; i++) {
    const x = x1 + incx * i;
    const y = y1 + incy * i;

    points.push([x, y]);
  }

  return points;
};

/**
 * Ecuacion de la recta que pasa por 2 puntos
 */
const line_equation_builder = (x1, y1, x2, y2) => {
  // pendiente
  const m = (y2 - y1) / (x2 - x1);
  // tÃ©rmino independiente
  const b = y1 - m * x1;

  const y = (x) => m * x + b;
  y.m = m;
  y.b = b;

  return y;
};

export { line_points, line_equation_builder };
