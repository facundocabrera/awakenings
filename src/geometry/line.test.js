const { abs, sqrt, pow, atan } = Math;

import { line_points, line_equation_builder } from "./line";

test("line_points", () => {
  const p2 = [1, 0];
  const p1 = [0, 1];
  const step = sqrt(2);

  const result = [
    p1,
    p2
  ]

  const current = line_points(...p1, ...p2, step);

  expect(
    current
  ).toEqual(
    result
  );
});

test("line_points", () => {
  const p1 = [-1, 0];
  const p2 = [0, -1];
  const step = sqrt(2);

  const result = [
    p1,
    p2
  ]

  const current = line_points(...p1, ...p2, step);

  expect(
    current
  ).toEqual(
    result
  );
});

test("line_points", () => {
  const p1 = [-2, 0];
  const p2 = [ 0, 2];
  const step = sqrt(2);

  const result = [
    p1,
    [-1, 1],
    p2
  ]

  const current = line_points(...p1, ...p2, step);

  expect(
    current
  ).toEqual(
    result
  );
});

test("line_points", () => {
  const p1 = [0, 3];
  const p2 = [3, 0];
  const step = sqrt(2);

  const result = [
    p1,
    [1, 2],
    [2, 1],
    p2
  ]

  const current = line_points(...p1, ...p2, step);

  console.log(...current);
  console.log(...result);

  expect(
    current
  ).toEqual(
    result
  );
});

test("line_equation_builder", () => {
  // [x, y] format
  const p1 = [0, 1];
  const p2 = [1, 0];
  const p3 = [-1, 2];

  const line_fn = line_equation_builder(...p1, ...p2);

  // line_fn(x) = y
  expect(line_fn(p1[0])).toBe(p1[1]);
  expect(line_fn(p2[0])).toBe(p2[1]);
  expect(line_fn(p3[0])).toBe(p3[1]);
});