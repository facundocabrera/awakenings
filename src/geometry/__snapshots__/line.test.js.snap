// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`line_points + stops(6) 1`] = `
Array [
  Array [
    1,
    0,
  ],
  Array [
    0.4999944999697543,
    0.8660095260523853,
  ],
]
`;

exports[`line_points: (-3, 0) => (0, 3) 1`] = `
Array [
  Array [
    -3,
    0,
  ],
  Array [
    -2,
    0.9999999999999998,
  ],
  Array [
    -1,
    1.9999999999999996,
  ],
  Array [
    0,
    2.999999999999999,
  ],
]
`;

exports[`line_points: (-4, 0) => (0, -4) 1`] = `
Array [
  Array [
    -4,
    0,
  ],
  Array [
    -3,
    -0.9999999999999998,
  ],
  Array [
    -2,
    -1.9999999999999996,
  ],
  Array [
    -1,
    -2.999999999999999,
  ],
  Array [
    0,
    -3.999999999999999,
  ],
]
`;

exports[`line_points: (0, -5) => (5, 0) 1`] = `
Array [
  Array [
    0,
    -5,
  ],
  Array [
    1,
    -4,
  ],
  Array [
    2,
    -3.0000000000000004,
  ],
  Array [
    3,
    -2.000000000000001,
  ],
  Array [
    4,
    -1.0000000000000009,
  ],
  Array [
    5,
    -8.881784197001252e-16,
  ],
]
`;

exports[`line_points: (0, 2) => (2, 0) 1`] = `
Array [
  Array [
    0,
    2,
  ],
  Array [
    1,
    1.0000000000000002,
  ],
  Array [
    2,
    4.440892098500626e-16,
  ],
]
`;

exports[`line_points: (0.5 * 2, 0.866 * 2) => (-0.5 * 2, 0.866 * 2) 1`] = `
Array [
  Array [
    1,
    1.732,
  ],
  Array [
    -0.000022000726026538686,
    1.732,
  ],
  Array [
    -1.000044001452053,
    1.732,
  ],
]
`;

exports[`line_points: (0.5, 0.866) => (-0.5, 0.866) 1`] = `
Array [
  Array [
    0.5,
    0.866,
  ],
  Array [
    -0.5000220007260265,
    0.866,
  ],
]
`;
